contracts/WardenHandler.sol

```diff
@@ -0,0 +1,311 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity =0.8.26;
+
+import '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+import '@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol';
+import '@openzeppelin/contracts/utils/Strings.sol';
+import './interfaces/Axelar/IAxelarGateway.sol';
+import './interfaces/Axelar/IAxelarGasService.sol';
+
+/// @notice Handles requests from Warden
+abstract contract WardenHandler is Initializable {
+  error NotApprovedByGateway();
+  error InvalidAddress();
+  error InvalidSourceChain();
+  error InvalidActionType();
+  error AmountTooBig();
+
+  uint8 private constant STAKE_ACTION_TYPE = 0;
+  uint8 private constant UUNSTAKE_ACTION_TYPE = 1;
+  uint8 private constant REINIT_ACTION_TYPE = 2;
+
+  uint8 internal constant SUCCESS_STATUS = 0;
+  uint8 internal constant ERROR_STATUS = 1;
+
+  /// keccak256(abi.encode(uint256(keccak256("eq-lab.storage.WardenHandlerData")) - 1)) & ~bytes32(uint256(0xff));
+  bytes32 private constant WardenHandlerDataStorageLocation =
+    0x4f376997038d6e5610d23f9f89ae844faaf6e156ed92caa3ff61a3cac093a900;
+
+  struct WardenHandlerData {
+    address axelarGateway;
+    address axelarGasService;
+    string wardenChain;
+    string wardenContractAddress;
+  }
+
+  struct WardenRequest {
+    uint8 actionType;
+    uint64 actionId;
+    uint128 lpAmount;
+  }
+
+  struct StakeResult {
+    uint8 status;
+    uint128 lpAmount;
+    uint128 tokenAmount;
```

Comment:
The field name is confusing, it's better to call it `reinitUnstakeTokenAmount`, imho